_posts/2018-10-03-ssl-logging-in-otp-22.md

```diff
@@ -0,0 +1,268 @@
+    01 00 00 9d 03 03 5b ac a1 cc 20 4c 4d 52 d0 d4
+    c8 fc dd 95 b0 fa 65 97 57 9e 44 aa dd 0e 46 10
+    6c 14 57 9c ce a0 00 00 04 00 ff c0 14 01 00 00
+    70 00 2b 00 06 00 04 03 04 03 03 00 00 00 0e 00
+    0c 00 00 09 6c 6f 63 61 6c 68 6f 73 74 00 0a 00
+    3a 00 38 00 0e 00 0d 00 19 00 1c 00 0b 00 0c 00
+    1b 00 18 00 09 00 0a 00 1a 00 16 00 17 00 08 00
+    06 00 07 00 14 00 15 00 04 00 05 00 12 00 13 00
+    01 00 02 00 03 00 0f 00 10 00 11 00 0b 00 02 01
+    00 00 32 00 04 00 02 02 03 00 0d 00 04 00 02 02
+    01
+...
+[Truncated for brevity]
+```
+
+The verbose debug logging proved to be especially useful during the development of new
+extensions as previously we had to use wireshark captures to validate TLS protocol
+messages.
+
+## Implementation
+
+In the ```ssl``` application we needed a way to handle two types of protocol
+messages (tls_record and handshake) each with a custom formatter function.
+
+The most straightforward solution was to add a new handler instance to the logger with a
+special formatter function that filters out all the "noise" coming from other
+modules of the system.
+
+The handler itself could be based on ```logger_std_h```, there was no need to implement
+a new one.
+
+```erlang
+logger:add_handler(ssl_handler, logger_std_h, Config),
+```
+
+The new ssl_handler is configured with a formatter that is implemented by the ```ssl_logger```
+module.
+
+```erlang
+Config = #{level => debug,
+           filter_default => stop,
+           formatter => {ssl_logger, #{}}},
+```
+
+Handler filter level is set to ```debug``` with ```stop``` as the default filter action. We also
+need a filter that lets the log events pass to the formatter if the source of the log event is the
+ssl application. In other words we need a domain filter with the action ```log``` on all sub-domains
+matching ```[otp,ssl]```.
+
+```erlang
+Filter = {fun logger_filters:domain/2,{log,sub,[otp,ssl]}},
+```
+
+Putting it all together we get the following function.
+
+```erlang
+start_logger() ->
+    Config = #{level => debug,
+               filter_default => stop,
+               formatter => {ssl_logger, #{}}},
+    Filter = {fun logger_filters:domain/2,{log,sub,[otp,ssl]}},
+    logger:add_handler(ssl_handler, logger_std_h, Config),
+    logger:add_handler_filter(ssl_handler, filter_non_ssl, Filter).
+```
+
+The function ```format``` is called in ssl_logger when an event gets through all the filters:
+
+```erlang
+format(#{level:= _Level, msg:= {report, Msg}, meta:= _Meta},
+       _Config0) ->
+     #{direction := Direction,
+       protocol := Protocol,
+       message := BinMsg0} = Msg,
+    case Protocol of
+        'tls_record' ->
+            BinMsg = lists:flatten(BinMsg0),
+            format_tls_record(Direction, BinMsg);
+        'handshake' ->
+            format_handshake(Direction, BinMsg0);
+        _Other ->
+            []
+    end.
+```
+
+There are two more helper functions that wrap around the logging macros:
```

Comment:
Why do you have these helpers?

Reply:
The main reason was to be able to set logging level per TLS session. I will update the text.

Reply:
aha!